test/util/test_case.hpp

```diff
@@ -28,63 +28,114 @@ namespace ngraph
+        /// NgraphTestCase
+        enum class BackendMode
+        {
+            // Use static version of backend
+            STATIC,
+            // Use dynamic version of backend
+            DYNAMIC
+        };
+
         class NgraphTestCase
         {
         public:
             NgraphTestCase(const std::shared_ptr<Function>& function,
-                           const std::string& backend_name)
-                : m_function(function)
-                , m_backend(ngraph::runtime::Backend::create(backend_name))
-            {
-            }
+                           const std::string& backend_name,
+                           BackendMode mode = BackendMode::DYNAMIC);
 
             /// \brief Makes the test case print the expected and computed values to the console.
             ///        This should only be used for debugging purposes.
             ///
             /// Just before the assertion is done, the current test case will gather expected and
             /// computed values, format them as 2 columns and print out to the console along with
-            //  a corresponding index in the vector.
+            /// a corresponding index in the vector.
             ///
             /// \param dump - Indicates if the test case should perform the console printout
             NgraphTestCase& dump_results(bool dump = true);
 
             template <typename T>
-            void add_input(const std::vector<T>& values)
+            void add_input(const Shape& shape, const std::vector<T>& values)
             {
-                auto params = m_function->get_parameters();
-
+                const auto params = m_function->get_parameters();
                 NGRAPH_CHECK(m_input_index < params.size(),
                              "All function parameters already have inputs.");
 
-                auto tensor = m_backend->create_tensor(params.at(m_input_index)->get_element_type(),
-                                                       params.at(m_input_index)->get_shape());
+                const auto& input_pshape = params.at(m_input_index)->get_partial_shape();
+                NGRAPH_CHECK(input_pshape.compatible(shape),
+                             "Passed input shape is not compatible with nGraph function.");
+
+                auto tensor =
+                    m_backend->create_tensor(params.at(m_input_index)->get_element_type(), shape);
                 copy_data(tensor, values);
 
                 m_input_tensors.push_back(tensor);
 
                 ++m_input_index;
             }
 
+            template <typename T>
+            void add_input(const std::vector<T>& values)
+            {
+                const auto& input_pshape =
+                    m_function->get_parameters().at(m_input_index)->get_partial_shape();
+                NGRAPH_CHECK(input_pshape.is_static(),
+                             "Input shape defined in Function must be fully known when static "
+                             "backend is used.");
```

Comment:
I think this comment may be misleading. You don't know and you're not checking at this moment whether you really deal with _dynamic_ backend. So IMHO the err message should be sth like: "Input data shape must be provided, if shape defined in Functions is not fully known."

Reply:
Sounds good for me